include/ofi_util.h

```diff
@@ -486,6 +486,75 @@ int ofi_get_src_addr(uint32_t addr_format,
 void ofi_getnodename(char *buf, int buflen);
 int ofi_av_get_index(struct util_av *av, const void *addr);
 
+/*
+ * Comamnd Queue
+ */
+
+struct util_cmd_handle {
+	struct dlist_entry entry;
+};
+
+typedef struct util_cmd_handle* (*util_cmd_queue_alloc)(size_t cmd_size);
+typedef void (*util_cmd_queue_free)(struct util_cmd_handle *handle);
+
+struct util_cmd_queue_attr {
+	util_cmd_queue_alloc cmd_alloc;
+	util_cmd_queue_free cmd_free;
+};
+
+struct util_cmd_queue {
+	struct dlist_entry queue;
+	struct util_cmd_queue_attr attr;
+	fastlock_t lock;
+};
+
+static inline
+int util_cmd_queue_empty(struct util_cmd_queue *queue)
+{
+	return dlist_empty(&queue->queue);
+}
+
+/* Retrieve a command from the head of CMD queue */
+static inline struct util_cmd_handle *
+util_cmd_queue_read(struct util_cmd_queue *queue)
+{
+	struct util_cmd_handle *handle;
+
+	if (util_cmd_queue_empty(queue))
+		return NULL;
+
+	fastlock_acquire(&queue->lock);
+	dlist_pop_front(&queue->queue, struct util_cmd_handle,
+			handle, entry);
+	fastlock_release(&queue->lock);
+
+	return handle;
+}
+
+/* Insert a command the tail of CMD queue */
+static inline void
+util_cmd_queue_write(struct util_cmd_queue *queue,
+		     struct util_cmd_handle *handle)
+{
+	fastlock_acquire(&queue->lock);
+	dlist_insert_tail(&handle->entry, &queue->queue);
+	fastlock_release(&queue->lock);
+}
+
+static inline struct util_cmd_handle *
+util_cmd_queue_alloc_cmd(struct util_cmd_queue *queue, size_t cmd_size)
+{
+	assert(cmd_size >= sizeof(struct util_cmd_handle));
+	return queue->attr.cmd_alloc(cmd_size);
+}
+
+static inline void
+util_cmd_queue_free_cmd(struct util_cmd_queue *queue,
+			struct util_cmd_handle *handle)
+{
+	queue->attr.cmd_free(handle);
+}
```

Comment:
This abstraction seems of questionable value.  The caller has to implement the alloc/free calls themselves.  The cmd_queue simply stores pointers to them, and then provides single line function wrappers around them.  The only real code it has is to queue/dequeue items into a double-linked list.

As an abstraction, we would be better off having a multi-threaded version of dlist.

Reply:
Yes, I agree. do you mean to have lock-less version of dlist/slist?

Reply:
The current version of dlist/slist are lockless.  I meant adding a lock-full? version.

Reply:
I'm not sure that this implementation is lockless, it rather lock-free, isn't it?

Reply:
The current dlist implementation is lockless already.  The cmd queue is basically a wrapper around dlist, but adds locking.  I'm suggesting that a better abstraction than cmd queue is a thread-safe version of dlist -- one that adds locking.  Note that I'm not proposing that we add locking to dlist directly.  I'm suggesting a new version, say dlist_mt, that has a better name and locks around the dlist calls.

Reply:
Ok, got it.